semeio/jobs/ots/ots.py

```diff
@@ -0,0 +1,597 @@
+        overburden timeshift calculations.
+        """
+        case = os.path.splitext(eclbase)[0]
+        self._init_file = EclFile("%s.INIT" % case)
+        self._rst_file = EclFile("%s.UNRST" % case)
+        self._grid = EclGrid("%s.EGRID" % case, apply_mapaxes=mapaxes)
+
+        self.subsidence = EclSubsidence(self._grid, self._init_file)
+
+        self._seabed = seabed
+        self._youngs_modulus = youngs * 1e9
+        self._poisson_ratio = poisson
+        self._r_factor = rfactor
+        self._convention = convention
+        self._verbose = verbose
+
+        self._surface = res_surface = OTSResSurface(grid=self._grid, above=above)
+        if velocity_model is not None:
+            self._surface = OTSVelSurface(res_surface=res_surface, vcube=velocity_model)
+
+        self._restart_views = {}
+
+    def get_horizon(self):
+        return self._create_surface()
+
+    def _create_surface(self, z=None):
+        """
+        Generate irap surface
+
+        :param z: replace z values of surface
+        """
+        nx = self._surface.nx
+        ny = self._surface.ny
+        x = self._surface.x
+        y = self._surface.y
+        if z is None:
+            z = self._surface.z
+
+        xstart = np.min(x)
+        ystart = np.min(y)
+
+        if nx < 2 or ny < 2:
+            raise RuntimeError("Cannot create IRAP surface if nx or ny is <2")
+
+        xinc = (np.max(x) - xstart) / (nx - 1)
+        yinc = (np.max(y) - ystart) / (ny - 1)
+
+        surf = Surface(
+            nx=nx, ny=ny, xinc=xinc, yinc=yinc, xstart=xstart, ystart=ystart, angle=0
+        )
+
+        irap_x = np.empty(nx * ny)
+        irap_y = np.array(irap_x)
+
+        for i, s in enumerate(surf):
+            irap_x[i], irap_y[i] = surf.getXY(i)
+
+        # Interpolate vel grid to irap grid, should be the same apart from ordering
+        z = np.nan_to_num(z)
+        ip = CloughTocher2DInterpolator((x, y), z, fill_value=0)
+        irap_z = ip(irap_x, irap_y)
+
+        for i, _ in enumerate(surf):
+            surf[i] = irap_z[i]
+
+        return surf
+
+    def add_survey(self, name, date):
+        """The add_survey() method will register a survey at a specific date.
+
+        The name argument should be a unique string, this will later
+        be used when evaluating the elastic strain in the
+        overburden. The date should be python datetime.date()
+        instance, this date should be present as a report step in the
+        restart file - otherwise an exception will be raised.
+        """
+        restart_view = self._rst_file.restartView(sim_time=date)
+        self.subsidence.add_survey_PRESSURE(name, restart_view)
+
+        self._restart_views[name] = restart_view
+
+        return restart_view
+
+    @staticmethod
+    def _divide_negative_shift(ts):
+        for i, _ in enumerate(ts):
+            if ts[i] < 0:
+                ts[i] /= 5.0
+
+    def geertsma_ts_rporv(self, vintage_pairs):
```

Comment:
The `geertsma_ts_rporv`and `geertsma_ts_simple` are identical except for the line `if len(vintage_pairs) < 1:`(0 in one of the function, 1 in the other), and the function they are calling (which could've been a function argument).

I woul recommend that those two functions are merged.

Reply:
If I would know more about details than perhaps it might be a good idea, but since the extracted description that is in CS schema says something I'm leaving it as is.

Reply:
The implementation should really not be effected by what the documentation says.

To be honest, I think it makes things much more transparent if we were to merge the functions, because then the differences between the implementations would be obvious. Instead of how it is now, where we need to diff the entire function to see what's different.

Reply:
> The geertsma_ts_rporvand geertsma_ts_simple are identical except for the line `if len(vintage_pairs) < 1`:(0 in one of the function, 1 in the other), and the function they are calling (which could've been a function argument).

A small brainfart here.. it's doing:

```
if len(vintage_pairs) < 1:
    return 0, []
```

Reply:
For the sake of wider coverage I took
```python
if len(vintage_pairs) < 1:
    return 0, []
```

Reply:
A suggestion at least.. I think I got most of it correct, haven't run it or anything. And even still the default arguments of the `delta_function` could be included higher. I am not completely sure how sane it is to do the partial on those functions, but I think I favor it. If we do it even further up we could make it even more obvious, that the defaults for poisson_ratio etc is the same all over. It's really just the survey names and position that changes.

Would want a better name for `du` as well. To be honest not sure what to call it.

```
    def delta_surface_calculation(self, vintage_pairs, delta_function=self.subsidence.evalGeertsma):
        """
        Must have good doc string here!
        """

        if len(vintage_pairs) == 0:
            # Always return same type
            return np.zeros((0, 0))

        vintages = self._vintages_name_date(vintage_pairs)
        surface = self._surface
        idx_to_calculate = [i for i in len(surface) if not np.isnan(surface.z[i])]

        du = np.zeros((len(vintages.name), len(surface)))
        ts_surfaces = []
        for vintage_index, (vintage_name, vintage_date) in enumerate(zip(vintages.name, vintages.date)):
            self.add_survey(vintage_name, vintage_date)

            dz_func = partial(delta_function(
                    base_survey=vintage_name,
                    monitor_survey=None,
                    youngs_modulus=self._youngs_modulus,
                    poisson_ratio=self._poisson_ratio,
                    seabed=self._seabed,
                )

            for idx in idx_to_calculate:
                r1 = (surface.x[idx], surface.y[idx], 0)
                r2 = (
                    surface.x[idx],
                    surface.y[idx],
                    surface.z[idx] - self._seabed,
                )
                # NB: comment below from pre refactor, not sure if this is well formulated

                # subsidence and displacement have opposite sign
                # should have minus on dz1 and dz2 here, more efficient
                # when calculating ts
                du[vintage_index, idx] = dz_func(pos=r1) - dz_func(pos=r2)

        for (base, monitor) in vintage_pairs:
            self._report(delta_function.__name__, base, monitor)

            base_index = vintages.date.index(base)
            monitor_index = vintages.date.index(monitor)
            ts = -self._r_factor * (du[monitor_index] - du[base_index])
            self._divide_negative_shift(ts)

            ts = ts * self._convention

            ts_surfaces.append(self._create_surface(ts))

        return ts_surfaces
```

Reply:
The usage of the subsidence function is so specific that I don't think having partial would bring any benefits neither the defaults for the matter. I renamed `du` to `surf_displacement` though.

Reply:
A bit hard to follow the discussion here, but what is the added benefit of calling `geertsma_ts_custom` from two different functions? From my point of view it is moving knowledge down in the call stack instead of up. I would rather split up `_gertsma_ts_custom` (which has a typo in the doc string, subsudence).

Reply:
I believe the current implementation now suffices for what I suggested at least.